dotnet/src/Planners/Planners.OpenAI/Stepwise/FunctionCallingStepwisePlanner.cs

```diff
@@ -25,52 +25,43 @@ public sealed class FunctionCallingStepwisePlanner
         this.Config = config ?? new();
-        this.Config.ExcludedPlugins.Add(RestrictedPluginName);
-
-        this._initialPlanPrompt = this.Config.GetPromptTemplate?.Invoke() ?? EmbeddedResource.Read("Stepwise.InitialPlanPrompt.txt");
         this._stepPrompt = this.Config.GetStepPromptTemplate?.Invoke() ?? EmbeddedResource.Read("Stepwise.StepPrompt.txt");
-
-        // Create context and logger
-        this._logger = loggerFactory.CreateLogger(this.GetType());
+        this.Config.ExcludedPlugins.Add(RestrictedPluginName);
     }
 
     /// <summary>
     /// Execute a plan
     /// </summary>
     /// <param name="question">The question to answer</param>
+    /// <param name="kernel">The <see cref="Kernel"/> containing services, plugins, and other state for use throughout the operation.</param>
     /// <param name="cancellationToken">The <see cref="CancellationToken"/> to monitor for cancellation requests. The default is <see cref="CancellationToken.None"/>.</param>
     /// <returns>Result containing the model's response message and chat history.</returns>
     public async Task<FunctionCallingStepwisePlannerResult> ExecuteAsync(
         string question,
+        Kernel kernel,
         CancellationToken cancellationToken = default)
     {
         Verify.NotNullOrWhiteSpace(question);
+        Verify.NotNull(kernel);
+        IChatCompletion chatCompletion = kernel.GetService<IChatCompletion>();
+        ILoggerFactory loggerFactory = kernel.GetService<ILoggerFactory>();
+        ILogger logger = loggerFactory.CreateLogger(this.GetType());
+        var promptTemplateFactory = new KernelPromptTemplateFactory(loggerFactory);
+        var stepExecutionSettings = this.Config.ExecutionSettings ?? new OpenAIPromptExecutionSettings();
 
-        // Add the final answer function
-        this._kernel.ImportPluginFromObject<UserInteraction>();
+        // Clone the kernel and modify it to add the final answer function
+        var clonedKernel = kernel.Clone();
+        clonedKernel.ImportPluginFromObject<UserInteraction>();
 
-        // Request completion for initial plan
-        var chatHistoryForPlan = await this.BuildChatHistoryForInitialPlanAsync(question, cancellationToken).ConfigureAwait(false);
-        string initialPlan = (await this._chatCompletion.GenerateMessageAsync(chatHistoryForPlan, null /* request settings */, this._kernel, cancellationToken).ConfigureAwait(false));
+        // Create and invoke a kernel function to generate the initial plan
+        var initialPlan = await this.GeneratePlanAsync(question, clonedKernel, logger, cancellationToken).ConfigureAwait(false);
 
-        var chatHistoryForSteps = await this.BuildChatHistoryForStepAsync(question, initialPlan, cancellationToken).ConfigureAwait(false);
+        var chatHistoryForSteps = await this.BuildChatHistoryForStepAsync(question, initialPlan, clonedKernel, chatCompletion, promptTemplateFactory, cancellationToken).ConfigureAwait(false);
```

Comment:
Is the GeneratePlan function still in the clonedKernel that is used for the executing the plan steps? If so should it be excluded from the function manual, or create a new cloned kernel?

Reply:
Good point, need to make sure the plugin name matches the excluded plugin name when registering this.